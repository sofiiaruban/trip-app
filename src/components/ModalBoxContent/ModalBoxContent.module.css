.content {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  padding: 0 24px 24px 24px;
}
.buttons-section {
  display: flex;
  justify-content: end;
  gap: 1rem;
}
.datePicker {
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  width: 100%;
}

.city {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}
