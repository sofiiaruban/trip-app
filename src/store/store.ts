import { configureStore } from '@reduxjs/toolkit'
import { setupListeners } from '@reduxjs/toolkit/query'
import { weatherDataApi } from '../services/weatherDataApi'
import tripReducer from './tripSlice'
import { persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import { combineReducers } from 'redux'

const persistConfig = {
  key: 'root',
  version: 1,
  storage
}

const reducer = combineReducers({
  [weatherDataApi.reducerPath]: weatherDataApi.reducer,
  trips: tripReducer
})
const persistedReducer = persistReducer(persistConfig, reducer)

export const store = configureStore({
  reducer: persistedReducer,

  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }).concat(weatherDataApi.middleware)
})

setupListeners(store.dispatch)
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
