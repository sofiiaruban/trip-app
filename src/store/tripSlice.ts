import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Trip } from '@app/types/types'
import initialTrips from '@app/initialTrips'

const initialState: Trip[] = initialTrips

const tripSlice = createSlice({
  name: 'trip',
  initialState,
  reducers: {
    addTrip(state, action: PayloadAction<Trip>) {
      state.push(action.payload)
    }
  }
})

export const { addTrip } = tripSlice.actions
export default tripSlice.reducer
