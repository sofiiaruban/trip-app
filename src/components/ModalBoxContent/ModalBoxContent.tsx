import { FC, FormEvent, useState } from 'react'
import Select from 'react-select'
import DatePicker from 'react-datepicker'
import 'react-datepicker/dist/react-datepicker.css'
import styles from './ModalBoxContent.module.css'
import Button from '../Buttons/Button'
import { ButtonColor, ButtonType } from '../Buttons/constants'
import { CityList } from '@app/constants'
import findImgUrlByCity from '@app/helpers/findImgUrlByCity'
import { addTrip } from '@app/store/tripSlice'
import { useDispatch } from 'react-redux'

interface Option {
  value: string
  label: string
}

const options: Option[] = [
  { value: CityList.KYIV, label: CityList.KYIV },
  { value: CityList.NEW_YORK, label: CityList.NEW_YORK },
  { value: CityList.LONDON, label: CityList.LONDON }
]

interface FormData {
  city: Option | null
  startDate: Date
  endDate: Date
}
interface ModalBoxContentProps {
  closeModal: () => void
}
const ModalBoxContent: FC<ModalBoxContentProps> = ({ closeModal }) => {
  const dispatch = useDispatch()

  const [formData, setFormData] = useState<FormData>({
    city: null,
    startDate: new Date(),
    endDate: new Date()
  })

  const handleInputChange = (fieldName: keyof FormData, value: unknown) => {
    setFormData((prevData) => ({
      ...prevData,
      [fieldName]: value
    }))
  }

  const handleSubmitForm = (event: FormEvent) => {
    event.preventDefault()
    if (formData.city) {
      const cityImgSrc = findImgUrlByCity(formData.city.value)
      if (cityImgSrc) {
        const tripData = {
          cityImgSrc,
          cityName: formData.city.label,
          startDate: formData.startDate!,
          endDate: formData.endDate!
        }
        dispatch(addTrip(tripData))
        closeModal()
      }
    }
  }

  return (
    <form onSubmit={handleSubmitForm} className={styles.content}>
      <div className={styles.city}>
        <label>
          <span>*</span> City:
        </label>
        <Select
          options={options}
          value={formData.city}
          onChange={(value) => handleInputChange('city', value)}
          placeholder="Please select city"
          required
        />
      </div>
      <div className={styles.datePicker}>
        <label>
          <span>*</span> Start date:
        </label>
        <DatePicker
          selected={formData.startDate}
          onChange={(date) => handleInputChange('startDate', date)}
          selectsStart
          startDate={formData.startDate}
          endDate={formData.endDate}
          placeholderText="Select date"
          required
        />
      </div>
      <div className={styles.datePicker}>
        <label>
          <span>*</span> End Date:
        </label>
        <DatePicker
          selected={formData.endDate}
          onChange={(date) => handleInputChange('endDate', date)}
          selectsEnd
          startDate={formData.startDate}
          endDate={formData.endDate}
          minDate={formData.startDate}
          placeholderText="Select date"
          required
        />
      </div>
      <div className={styles['buttons-section']}>
        <Button
          type={ButtonType.RESET}
          title="Cancel"
          color={ButtonColor.BASIC}
          onClick={closeModal}
        />
        <Button title="Save" />
      </div>
    </form>
  )
}

export default ModalBoxContent
