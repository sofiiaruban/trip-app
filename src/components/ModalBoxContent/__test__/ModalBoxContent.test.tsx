import { render } from '@testing-library/react'
import ModalBoxContent from '../ModalBoxContent'
import { store } from '@app/store/store'
import { Provider } from 'react-redux'

jest.mock('@reduxjs/toolkit/query/react', () => ({
  ...jest.requireActual('@reduxjs/toolkit/query/react'),
  fetchBaseQuery: jest.fn()
}))

test('ModalBoxContent matches snapshot', () => {
  const closeModalMock = jest.fn()
  const { container } = render(
    <Provider store={store}>
      <ModalBoxContent closeModal={closeModalMock} />
    </Provider>
  )

  expect(container).toMatchSnapshot()
})
