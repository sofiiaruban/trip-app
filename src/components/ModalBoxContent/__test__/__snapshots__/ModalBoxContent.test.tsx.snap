// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalBoxContent matches snapshot 1`] = `
<div>
  <form
    class="content"
  >
    <div
      class="city"
    >
      <label>
        <span>
          *
        </span>
         City:
      </label>
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class=" css-13cymwt-control"
        >
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            >
              Please select city
            </div>
            <div
              class=" css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="datePicker"
    >
      <label>
        <span>
          *
        </span>
         Start date:
      </label>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container react-datepicker__view-calendar-icon"
        >
          <svg
            class=" react-datepicker__calendar-icon undefined false"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="Calendar"
            >
              <path
                d="M18.438,4.954H16.5c0-0.346,0-0.691,0-1.036c0-0.124,0-0.248,0-0.372c0-0.262-0.23-0.512-0.5-0.5
            c-0.271,0.012-0.5,0.22-0.5,0.5c0,0.469,0,0.939,0,1.408h-7c0-0.346,0-0.691,0-1.036c0-0.124,0-0.248,0-0.372
            c0-0.262-0.23-0.512-0.5-0.5c-0.271,0.012-0.5,0.22-0.5,0.5c0,0.469,0,0.939,0,1.408H5.562c-1.378,0-2.5,1.122-2.5,2.5v11
            c0,1.379,1.122,2.5,2.5,2.5h12.875c1.379,0,2.5-1.121,2.5-2.5v-11C20.938,6.076,19.816,4.954,18.438,4.954z M5.562,5.954H7.5
            c0,0.073,0,0.147,0,0.22c0,0.124,0,0.248,0,0.372c0,0.262,0.23,0.512,0.5,0.5c0.271-0.012,0.5-0.22,0.5-0.5c0-0.197,0-0.394,0-0.592
            h7c0,0.073,0,0.147,0,0.22c0,0.124,0,0.248,0,0.372c0,0.262,0.23,0.512,0.5,0.5c0.271-0.012,0.5-0.22,0.5-0.5
            c0-0.197,0-0.394,0-0.592h1.937c0.827,0,1.5,0.673,1.5,1.5v1.584H4.062V7.454C4.062,6.627,4.735,5.954,5.562,5.954z M18.438,19.954
            H5.562c-0.827,0-1.5-0.673-1.5-1.5v-8.416h15.875v8.416C19.938,19.281,19.265,19.954,18.438,19.954z"
              />
            </g>
          </svg>
          <input
            class=""
            placeholder="Select date"
            type="text"
            value="02/23/2024"
          />
        </div>
      </div>
    </div>
    <div
      class="datePicker"
    >
      <label>
        <span>
          *
        </span>
         End Date:
      </label>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container react-datepicker__view-calendar-icon"
        >
          <svg
            class=" react-datepicker__calendar-icon undefined false"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="Calendar"
            >
              <path
                d="M18.438,4.954H16.5c0-0.346,0-0.691,0-1.036c0-0.124,0-0.248,0-0.372c0-0.262-0.23-0.512-0.5-0.5
            c-0.271,0.012-0.5,0.22-0.5,0.5c0,0.469,0,0.939,0,1.408h-7c0-0.346,0-0.691,0-1.036c0-0.124,0-0.248,0-0.372
            c0-0.262-0.23-0.512-0.5-0.5c-0.271,0.012-0.5,0.22-0.5,0.5c0,0.469,0,0.939,0,1.408H5.562c-1.378,0-2.5,1.122-2.5,2.5v11
            c0,1.379,1.122,2.5,2.5,2.5h12.875c1.379,0,2.5-1.121,2.5-2.5v-11C20.938,6.076,19.816,4.954,18.438,4.954z M5.562,5.954H7.5
            c0,0.073,0,0.147,0,0.22c0,0.124,0,0.248,0,0.372c0,0.262,0.23,0.512,0.5,0.5c0.271-0.012,0.5-0.22,0.5-0.5c0-0.197,0-0.394,0-0.592
            h7c0,0.073,0,0.147,0,0.22c0,0.124,0,0.248,0,0.372c0,0.262,0.23,0.512,0.5,0.5c0.271-0.012,0.5-0.22,0.5-0.5
            c0-0.197,0-0.394,0-0.592h1.937c0.827,0,1.5,0.673,1.5,1.5v1.584H4.062V7.454C4.062,6.627,4.735,5.954,5.562,5.954z M18.438,19.954
            H5.562c-0.827,0-1.5-0.673-1.5-1.5v-8.416h15.875v8.416C19.938,19.281,19.265,19.954,18.438,19.954z"
              />
            </g>
          </svg>
          <input
            class=""
            placeholder="Select date"
            type="text"
            value="02/23/2024"
          />
        </div>
      </div>
    </div>
    <div
      class="buttons-section"
    >
      <button
        class="button basic"
        type="reset"
      >
        Cancel
      </button>
      <button
        class="button primary"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
